Remove global ProcessPoolExecutor

cburgdorf opened the pull request:
### What was wrong?

Trinity does currently leave processes behind. Read #897 for a more detailed description.

I used the following steps to reproduce the issue:

1. Check you have no dangling processes `ps aux |grep trinity`
2. If you do, kill them. (`pkill python` if you are brave)
3. Run beam sync (*should* work with other sync, too but I used that) until it is fully involved syncing
4. ☝ This is important, killing Trinity when it isn't yet syncing does not leave these processes behind.
5. Kill trinity with `CTRL + C`

Another important hint: One can check the process ids with these of the plugin processes. It makes it clear that those process ids are not ours. They really are from the `ProcessPoolExecutor`

### How was it fixed?

I've spent almost the entire day on this and it got me so frustrated that the outcome now is this removal of the entire `ProcessPoolExecutor`.

Here's the backstory:

1. The problem of leftover processes was originally discovered by @veox and back then it was even more severe as every isolated plugin would create a ton of idle processes.

2. I fixed this back then in https://github.com/ethereum/trinity/pull/227

3. Back then, @pipermerriam already [raised concerns](https://github.com/ethereum/trinity/pull/227#pullrequestreview-197099468) with the solution, noting that someone may accidentally remove it at some point. He also proposed using a `weakref.finalizer` for a more robust solution.

4. Ironically it is @pipermerriam 's PR under my review that then happens to do exactly that: [accidentally remove the fix](https://github.com/ethereum/trinity/commit/6493c94656b907ca9f73ce2fe677dfd9f03a2974#diff-31de14ce34c9eb0bd0c3a7fed15bda1eL141).

Now, I've spend a fair amount of time trying to come up with something more robust including the proposed `weakref.finalizer` solution. Unfortunately, it has proven to be really hard (or maybe that's just me) to come up with something that works reliably.

This got me questioning if we really need the `ProcessPoolExecutor` here or if a simple `ThreadPoolExecutor` might do it as well.

I haven't done any profiling about the performance impact of this.

### To-Do

[//]: # (Stay ahead of things, add list items here!)
- [ ] Clean up commit history

[//]: # (For important changes that should go into the release notes please add a newsfragment file as explained here: https://github.com/ethereum/trinity/blob/master/newsfragments/README.md)

[//]: # (See: https://trinity-client.readthedocs.io/en/latest/contributing.html#pull-requests)
- [ ] Add entry to the [release notes](https://github.com/ethereum/trinity/blob/master/newsfragments/README.md)

#### Cute Animal Picture

![put a cute animal picture link inside the parentheses](https://static.boredpanda.com/blog/wp-content/uploads/2014/03/cute-smiling-animals-9.jpg)

pipermerriam commented:
I'm :+1: on trying this approach if the performance comparison is comprable.  I'll leave it to @carver to evaluate as he's the defacto "is this ok for beam sync" person.

carver commented:
:x: for CAP ;)

cburgdorf commented:
>For Beam Sync, every method that needs a separate process is in its own isolated process plugin, AFAICT.

Correct. And it's also the case that when the decision for the `ProcessPoolExecutor` was made, there were many things that were running in the same process that are running in their own processes today anyway. (request server, discovery, upnp, tx pool etc)

 
>:x: for CAP X)

Fixed that and added a release notes entry :+1:

carver commented:
Confirmed, fixes #897
